// slash-commands/dist/src/structures/MessageFlags.js
"use strict";
// JSDoc comments taken from https://discord.com/developers/docs/resources/channel#message-object-message-flags
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageFlags = void 0;
var MessageFlags;
(function (MessageFlags) {
    /** This message has been published to subscribed channels (via Channel Following) */
    MessageFlags[MessageFlags["CROSSPOSTED"] = 1] = "CROSSPOSTED";
    /** This message originated from a message in another channel (via Channel Following) */
    MessageFlags[MessageFlags["IS_CROSSPOST"] = 2] = "IS_CROSSPOST";
    /** Do not include any embeds when serializing this message */
    MessageFlags[MessageFlags["SUPPRESS_EMBEDS"] = 4] = "SUPPRESS_EMBEDS";
    /** The source message for this crosspost has been deleted (via Channel Following) */
    MessageFlags[MessageFlags["SOURCE_MESSAGE_DELETED"] = 8] = "SOURCE_MESSAGE_DELETED";
    /**	This message came from the urgent message system */
    MessageFlags[MessageFlags["URGENT"] = 16] = "URGENT";
    /** This message is only visible to the user who ran the command, similar to clyde */
    MessageFlags[MessageFlags["EPHEMERAL"] = 64] = "EPHEMERAL";
})(MessageFlags = exports.MessageFlags || (exports.MessageFlags = {}));
